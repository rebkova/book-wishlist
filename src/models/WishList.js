import { types } from 'mobx-state-tree'

// this piece of data is called a snapshot
const data = {
  "title": "Bryson's Dictionary of Troublesome Words: A Writer's Guide to Getting It Right",
  "authors": "Bill Bryson",
  "price": 3.88,
}

export const WishListItem = types.model({
  title: types.string,
  authors: types.string,
  price: types.number,
})
  .actions(self => ({
    // using function with 'self' parameter to avoid issues with 'this' keyword
    changeTitle(newTitle) {
      self.title = newTitle
    },
    changePrice(newPrice) {
      self.price = newPrice
    },
    changeAuthors(newAuthor) {
      self.authors = newAuthor
    },
  }))

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => ({
    add(item) {
      self.items.push(item)
    }
  }))