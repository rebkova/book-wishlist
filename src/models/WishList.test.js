import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { WishList, WishListItem } from "./WishList";

it('can create an instance of a model', () => {
  const item = WishListItem.create({
    title: "Bryson's Dictionary of Troublesome Words: A Writer's Guide to Getting It Right",
    authors: "Bill Bryson",
    price: 3.88,
  })

  expect(item.price).toBe(3.88)
  item.changeTitle('Baloo')
  expect(item.title).toBe('Baloo')
})

it('can create a wishlist', () => {
  const list = WishList.create({
    items: [
      {
        title: "Bryson's Dictionary of Troublesome Words: A Writer's Guide to Getting It Right",
        authors: "Bill Bryson",
        price: 3.88,
      },
    ]
  })

  expect(list.items.length).toBe(1)
  expect(list.items[0].price).toBe(3.88)
})

it('can add new items', () => {
  const list = WishList.create()
  const states = []
  onSnapshot(list, snapshot => {
    states.push(snapshot)
  })

  list.add(WishListItem.create({
    title: "Pika Nogavička",
    authors: "Astrid Lindgren",
    price: 4.25,
  }))

  expect(list.items.length).toBe(1)
  expect(list.items[0].title).toBe('Pika Nogavička')
  list.items[0].changeTitle('Pippi!')
  expect(list.items[0].title).toBe('Pippi!')

  // instead of writing tests for every property, we compare
  // the whole model to itself
  expect(getSnapshot(list)).toMatchSnapshot()
  expect(states).toMatchSnapshot()
})

it('can add new items - 2', () => {
  const list = WishList.create()
  const patches = []
  onPatch(list, patch => {
    patches.push(patch)
  })

  list.add(WishListItem.create({
    title: "Pika Nogavička",
    authors: "Astrid Lindgren",
    price: 4.25,
  }))

  list.items[0].changeTitle('Emil i Löneberga');
  expect(list.items[0].title).toBe('Emil i Löneberga');

  // expect(getSnapshot(list)).toMatchSnapshot();

  expect(onPatch(patches)).toMatchSnapshot();
})