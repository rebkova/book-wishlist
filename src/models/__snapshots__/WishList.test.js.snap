// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "authors": "Astrid Lindgren",
      "price": 4.25,
      "title": "Pika Nogavička",
    },
  },
]
`;

exports[`can add new items - 2 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "authors": "Astrid Lindgren",
      "price": 4.25,
      "title": "Pika Nogavička",
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/title",
    "value": "Emil i Löneberga",
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "authors": "Astrid Lindgren",
      "price": 4.25,
      "title": "Pippi!",
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "authors": "Astrid Lindgren",
        "price": 4.25,
        "title": "Pika Nogavička",
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "authors": "Astrid Lindgren",
        "price": 4.25,
        "title": "Pippi!",
      },
    ],
  },
]
`;
